File: М11_ClusterVMS/clustervms/cluster/publish.py
"""Lesson 3 — the one-way publication upward, and what the operator is told.

    object first, then the Variable (by CAS) — a pointer never dangles
    publish on change, with a floor — the RPO is the floor
    acknowledge on local commit, and SHOW durability: `replicated` condition

The directory is each Node's off-box backup. It never writes back.
"""
from __future__ import annotations

import logging
import time

from .variables import Conflict, Variables

log = logging.getLogger("cluster.publish")


class Publisher:
    def __init__(self, node: str, store, vars_: Variables, objects, floor_seconds: float = 5.0,
                 clock=time.monotonic):
        self.node, self.store, self.vars, self.objects = node, store, vars_, objects
        self.floor, self.clock = floor_seconds, clock
        self.published_rev = 0
        self.last_publish = -1e9
        self.last_attempt_failed_since: float | None = None
        self.publishes = 0

    async def publish_once(self) -> bool:
        """Publish if the local revision moved past what the directory holds.
        Returns True if a publish happened (or nothing was needed)."""
        rev = await self.store.config_revision()
        if rev <= self.published_rev:
            return True
        if self.clock() - self.last_publish < self.floor:
            return True                                   # the floor: not yet
        try:
            blob = await self.store.dump_config()
            key = f"{self.node}/rev-{rev}"
            self.objects.put(key, blob)                   # 1. the object, first
            items, idx = self.vars.get(f"nodes/{self.node}")
            items = dict(items or {})
            items.update({"node": self.node, "config": key, "revision": str(rev),
                          "cameras": ",".join(str(c["id"]) for c in await self.store.cameras())})
            self.vars.put(f"nodes/{self.node}", items, cas=idx)   # 2. then the pointer
        except Conflict:
            log.warning("publish: cas conflict on nodes/%s — somebody else wrote our Variable; retrying next tick", self.node)
            return False
        except Exception as e:                            # noqa: BLE001 — the store or the cluster is unreachable
            if self.last_attempt_failed_since is None:
                self.last_attempt_failed_since = self.clock()
            log.warning("publish failed (%s); the Node keeps running, the edit stays local", type(e).__name__)
            return False
        self.published_rev = rev
        self.last_publish = self.clock()
        self.last_attempt_failed_since = None
        self.publishes += 1
        return True

    def replicated(self, local_rev: int) -> tuple[bool, str | None]:
        """The `replicated` condition for the console: status and reason."""
        if local_rev <= self.published_rev:
            return True, None
        if self.last_attempt_failed_since is not None:
            return False, f"not yet replicated: directory unreachable for {self.clock() - self.last_attempt_failed_since:.0f}s"
        return False, f"not yet replicated ({local_rev - self.published_rev} edit(s) pending)"

File: М11_ClusterVMS/clustervms-go/cluster/publish.go
package cluster

// Lesson 3 — the one-way publication upward, and what the operator is told.
//
//	object first, then the Variable (by CAS) — a pointer never dangles
//	publish on change, with a floor — the RPO is the floor
//	acknowledge on local commit, and SHOW durability: the `replicated` condition
//
// The directory is each Node's off-box backup. It never writes back.

import (
	"errors"
	"fmt"
	"log"
	"strconv"
	"strings"
)

type Publisher struct {
	Node    string
	Store   Store
	Vars    Variables
	Objects ObjectStore
	Floor   float64
	Clock   Clock

	PublishedRev int64
	LastPublish  float64
	FailedSince  float64 // < 0: the last attempt succeeded
	Publishes    int
}

func NewPublisher(node string, store Store, v Variables, objects ObjectStore, floor float64, clock Clock) *Publisher {
	return &Publisher{Node: node, Store: store, Vars: v, Objects: objects, Floor: floor, Clock: clock,
		LastPublish: -1e9, FailedSince: -1}
}

// PublishOnce publishes if the local revision moved past what the directory
// holds. True if a publish happened or nothing was needed; false if it was
// attempted and failed (the Node keeps running, the edit stays local).
func (p *Publisher) PublishOnce() bool {
	rev, err := p.Store.ConfigRevision()
	if err != nil {
		return p.failed(err)
	}
	if rev <= p.PublishedRev {
		return true
	}
	if p.Clock()-p.LastPublish < p.Floor {
		return true // the floor: not yet
	}
	blob, err := p.Store.DumpConfig()
	if err != nil {
		return p.failed(err)
	}
	key := fmt.Sprintf("%s/rev-%d", p.Node, rev)
	if err := p.Objects.Put(key, blob); err != nil { // 1. the object, first
		return p.failed(err)
	}
	items, idx, err := p.Vars.Get("nodes/" + p.Node)
	if err != nil {
		return p.failed(err)
	}
	if items == nil {
		items = Items{}
	}
	cams, err := p.Store.Cameras()
	if err != nil {
		return p.failed(err)
	}
	ids := make([]string, 0, len(cams))
	for _, c := range cams {
		ids = append(ids, strconv.FormatInt(c.ID, 10))
	}
	items["node"], items["config"] = p.Node, key
	items["revision"], items["cameras"] = strconv.FormatInt(rev, 10), strings.Join(ids, ",")
	if _, err := p.Vars.Put("nodes/"+p.Node, items, idx); err != nil { // 2. then the pointer
		if errors.Is(err, ErrConflict) {
			log.Printf("publish: cas conflict on nodes/%s — somebody else wrote our Variable; retrying next tick", p.Node)
			return false
		}
		return p.failed(err)
	}
	p.PublishedRev, p.LastPublish, p.FailedSince = rev, p.Clock(), -1
	p.Publishes++
	return true
}

func (p *Publisher) failed(err error) bool {
	if p.FailedSince < 0 {
		p.FailedSince = p.Clock()
	}
	log.Printf("publish failed (%v); the Node keeps running, the edit stays local", err)
	return false
}

// Replicated is the `replicated` condition for the console: status and reason.
func (p *Publisher) Replicated(localRev int64) (bool, string) {
	if localRev <= p.PublishedRev {
		return true, ""
	}
	if p.FailedSince >= 0 {
		return false, fmt.Sprintf("not yet replicated: directory unreachable for %.0fs", p.Clock()-p.FailedSince)
	}
	return false, fmt.Sprintf("not yet replicated (%d edit(s) pending)", localRev-p.PublishedRev)
}
